src/VisualStudio/Core/Def/Implementation/DebuggerIntelliSense/DebuggerIntellisenseFilter.cs

```diff
@@ -17,26 +18,70 @@
 
 namespace Microsoft.VisualStudio.LanguageServices.Implementation.DebuggerIntelliSense
 {
-    internal class DebuggerIntelliSenseFilter<TPackage, TLanguageService> : AbstractVsTextViewFilter<TPackage, TLanguageService>, IDisposable
+    internal class DebuggerIntelliSenseFilter<TPackage, TLanguageService> : AbstractVsTextViewFilter<TPackage, TLanguageService>, IDisposable, IFeatureController
         where TPackage : AbstractPackage<TPackage, TLanguageService>
         where TLanguageService : AbstractLanguageService<TPackage, TLanguageService>
     {
         private readonly ICommandHandlerServiceFactory _commandFactory;
-        private readonly IWpfTextView _wpfTextView;
+        private readonly IFeatureServiceFactory _featureServiceFactory;
         private AbstractDebuggerIntelliSenseContext _context;
         private IOleCommandTarget _originalNextCommandFilter;
+        private IFeatureDisableToken _completionDisabledToken;
+        private object _completionDisabledTokenLock = new object();
 
         internal bool Enabled { get; set; }
 
         public DebuggerIntelliSenseFilter(
             AbstractLanguageService<TPackage, TLanguageService> languageService,
             IWpfTextView wpfTextView,
             IVsEditorAdaptersFactoryService adapterFactory,
-            ICommandHandlerServiceFactory commandFactory)
+            ICommandHandlerServiceFactory commandFactory,
+            IFeatureServiceFactory featureServiceFactory)
             : base(languageService, wpfTextView, adapterFactory, commandFactory)
         {
-            _wpfTextView = wpfTextView;
             _commandFactory = commandFactory;
+            _featureServiceFactory = featureServiceFactory;
+        }
+
+        internal void EnableCompletion()
+        {
+            var featureService = _featureServiceFactory.GetOrCreate(WpfTextView);
+            if (!featureService.IsEnabled(PredefinedEditorFeatureNames.Completion))
+            {
+                lock (_completionDisabledTokenLock)
```

Comment:
Why the synchronization? I would have imagined we only call this on the UI thread?

Reply:
I want to believe that all calls are performed sequentially. However, I do not see any checks for the UI thread on our side in this API, and have no guarantees that this actually happens. Should we just hope so and remove extra checks?

---
In reply to: [270162103](https://github.com/dotnet/roslyn/pull/32631#discussion_r270162103) [](ancestors = 270162103)

Reply:
This entire class is UI affinitized -- no two methods can run at the same time. The other fields are all mutable and have no guards either.

If you want to actually put foreground assserts I wouldn't complain.

Reply:
I could not see an elegant way to get a threading context to DebuggerIntelliSenseFilter or to AbstractLanguageService<TPackage, TLanguageService> that could allow to add asserts. If you know, please let me know.

Removing locks.

---
In reply to: [270217422](https://github.com/dotnet/roslyn/pull/32631#discussion_r270217422) [](ancestors = 270217422)